### PR TITLE
[JENKINS-65195, JENKINS-64347] Fixes to ProcessTree implementation on Darwin

### PR DESCRIPTION
**Pull request by ngg**
See [JENKINS-65195](https://issues.jenkins-ci.org/browse/JENKINS-65195).
See [JENKINS-64347](https://issues.jenkins-ci.org/browse/JENKINS-64347).

### Proposed changelog entries

* JENKINS-65195 Fix incorrect process termination issues when running on macOS.
* JENKINS-64347 Add check to prevent out of bounds memory access on macOS. 

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@daniel-beck 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by timja** [comment]
Could you explain any testing / reproduction steps to help with verifying this?

Is there a consistent way to re-produce it?

I have apple arm64 on big sur so I can verify it myself if I know how.

Can't really comment on the code itself other than somethings changed from an int to a long so I guess the api changed in certain places, but do you have more information other than 'Fix incorrect process termination issues when running on macOS.' ?

The issue describes the symptoms but not why your fix is required

Thanks!

**Comment by car-roll** [comment]
@timja just my two cents: For reproduction, I think this issue is what is affecting  `durable-task`, specifically in [`BourneShellScriptTest#stop`](https://github.com/jenkinsci/durable-task-plugin/blob/d7b8e287d5964bd33e157ad456e31e6ec7b988be/src/test/java/org/jenkinsci/plugins/durabletask/BourneShellScriptTest.java#L243)
The test will fail under `BigSur` but succeed in older versions. It also passes under the suite of linux platforms.

**Comment by timja** [comment]
> @timja just my two cents: For reproduction, I think this issue is what is affecting `durable-task`, specifically in [`BourneShellScriptTest#stop`](https://github.com/jenkinsci/durable-task-plugin/blob/d7b8e287d5964bd33e157ad456e31e6ec7b988be/src/test/java/org/jenkinsci/plugins/durabletask/BourneShellScriptTest.java#L243)
> The test will fail under `BigSur` but succeed in older versions. It also passes under the suite of linux platforms.

@car-roll any thoughts on the below?

Test environment:

<table>
<tr>
	<td>
	<td>2.296
	<td>PR#5548
<tr>
	<td>Big Sur arm64
	<td>Failed (runs infinitely)
	<td>Passes (15min)
<tr>
	<td>Catalina x64
	<td>Passes (10sec)
	<td>Passes (14sec)
</table>

All OS tests except native fail when I run on my machine, maybe some setup I have? It does try run a docker container

The test takes 15minutes to pass after the fix, when I pause the test it's reading bytes from a fileinputstream

**Comment by ngg** [comment]
The changes are not ARM or Big Sur specific, these were bugs on older macOS versions on x64 as well.
There were two bugs that this PR fixes.

**First bug:**

The sysctl, sysctlbyname, sysctlnametomib functions were defined incorrectly, there were both 4-byte int vs 8-byte int differences and also pointer vs value differences.

The C definitions can be found here: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/sysctl.3.html)
```
int sysctl(int *name, u_int namelen, void *oldp, size_t *oldlenp, void *newp, size_t newlen);
int sysctlbyname(const char *name, void *oldp, size_t *oldlenp, void *newp, size_t newlen);
int sysctlnametomib(const char *name, int *mibp, size_t *sizep);
```

These were corrected in `GNUCLibrary.java`. The older definitions were marked as deprecated instead of being removed because there might be plugins that use these.

**Second bug:**

The `KERN_PROCARGS2` sysctl returns the number of bytes it has written in `*oldlenp`, but that was not used at all. A much larger memory chunk was given to the sysctl than it actually needs, which was not initialized to null, so it could contain any garbage value. The size was not checked properly, so parsing the arguments and environment variables were based on partly uninitialized memory.

This second bug caused the actual issue in [JENKINS-65195](https://issues.jenkins-ci.org/browse/JENKINS-65195) because processes environment variables were messed up and the process killer finds which processes to kill based on the envvars.

**Comment by timja** [comment]
> The changes are not ARM or Big Sur specific, these were bugs on older macOS versions on x64 as well.
There were two bugs that this PR fixes.

Is it interesting that it fixes the test https://github.com/jenkinsci/durable-task-plugin/blob/d7b8e287d5964bd33e157ad456e31e6ec7b988be/src/test/java/org/jenkinsci/plugins/durabletask/BourneShellScriptTest.java#L243 then on Big Sur? (Although it takes a long time to pass for some reason)

**Comment by ngg** [comment]
> Is it interesting that it fixes the test https://github.com/jenkinsci/durable-task-plugin/blob/d7b8e287d5964bd33e157ad456e31e6ec7b988be/src/test/java/org/jenkinsci/plugins/durabletask/BourneShellScriptTest.java#L243 then on Big Sur? (Although it takes a long time to pass for some reason)

I think the environment (OS version included) affect what those uninitialized memory chunks contain, so it can make bugs occur more frequently. When we first found [JENKINS-65195](https://issues.jenkins-ci.org/browse/JENKINS-65195) we found it on ARM Big Sur, but later the bug also appeared on x64 as well, only much less frequently.

File: core/src/main/java/hudson/util/ProcessTree.java
Comment (res0nance) [inline core/src/main/java/hudson/util/ProcessTree.java:1783]: This change should solve https://issues.jenkins.io/browse/JENKINS-64347

File: core/src/main/java/hudson/util/ProcessTree.java
Comment (res0nance) [inline core/src/main/java/hudson/util/ProcessTree.java:1799]: Technically doesn't matter since the memory has been zero initialized but might be useful for debugging

**Review by res0nance — COMMENTED** [review]
Now personally I'm not sure how this solves the issue because considering if `argc` holds the arguments and is successfully parsed. The issue comes with parsing everything after.

Does that mean that the memory does not hold null terminated strings?

**Comment by res0nance** [comment]
Thinking out loud here, what if `argc` is 0? won't Skip0 skip it? What are we even reading then?

File: core/src/main/java/hudson/util/ProcessTree.java
Comment (timja) [reply core/src/main/java/hudson/util/ProcessTree.java:1783]: I added it to the changelog above

**Comment by basil** [comment]
For comparison, there is [another implementation of this in OSHI](https://github.com/oshi/oshi/blob/596d27a91f05d0393a44909e83ef244c0f654c1f/oshi-core/src/main/java/oshi/software/os/mac/MacOSProcess.java).

**Review by basil — APPROVED** [review]
The logic changes look correct to me. The function definition changes also look correct to me for both 32-bit macOS and 64-bit macOS, since [based on this page](https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/64bitPorting/transition/transition.html) `size_t` and `long` have the same lengths (4 bytes on 32-bit macOS and 8 bytes on 64-bit macOS).

**Comment by basil** [comment]
> For comparison, there is [another implementation of this in OSHI](https://github.com/oshi/oshi/blob/596d27a91f05d0393a44909e83ef244c0f654c1f/oshi-core/src/main/java/oshi/software/os/mac/MacOSProcess.java).

Coincidentally, a similar fix to OSHI's function definitions was merged 6 days ago in oshi/oshi#1649.

**Comment by basil** [comment]
Perhaps @dbwiddis, who has clearly been down this road before, might be interested in reviewing this PR.

**Comment by basil** [comment]
@ngg Out of curiosity, has this PR been tested on 32-bit macOS?

**Comment by dbwiddis** [comment]
> Perhaps @dbwiddis, who has clearly been down this road before, might be interested in reviewing this PR.

Two comments:
1. It's perfectly fair to blame the me in 2015 that submitted my first PR to the JNA project and messed up the type mapping.  I'm sorry.
2. While *most* of the time the byte sizes of `NativeLong` and `size_t` will match, it is not a guarantee and the proposed fix is not correct.  (I'll caveat that with saying it might be correct for macOS but that would be an implementation detail rather than documentation.)

I see you've already linked my temporary fix in OSHI in https://github.com/oshi/oshi/pull/1649

I've also submitted an upstream fix in JNA (targeted for 5.9.0). See https://github.com/java-native-access/jna/pull/1351

I'd recommend going the route in the JNA fix using temporary subclasses implementing those changes, as it'll be easier to remove them and use JNA 5.9.0 behavior once it's released.  Or, ask on the jna-users mailing list about the possibility of a release so you don't need an intermediate step.

**Comment by dbwiddis** [comment]
A third comment after re-reading this thread. If the assumption is going to be always 64 bit, then `NativeLong`, `size_t`, and `long` (and `LongByReference`) all would behave the same.  Among those choices:
 - `NativeLong` is not correct and introduces extra object creation vs. `long`
 - `long` (and `LongByReference`) would work for 64-bit only and in theory be slightly more efficient
 - `size_t` properly mapped is the "correct" solution but incurs a bit of extra object creation.

**Comment by basil** [comment]
Thank you for reviewing this, @dbwiddis! That's the same conclusion I came to: `NativeLong` and `NativeLongByReference` are technically incorrect but happen to always work for (64-bit?) macOS, which is the only environment in which this code needs to work in practice. I also agree that the correct solution would be to wait for JNA 5.9.0, add `jna-platform` to the build, and then use `size_t` and `size_t.ByReference`. But given how many Jenkins users this has affected, I would not want to block a working solution waiting for a perfect one. That's why I approved this PR. 😊

BTW I'm a big fan of OSHI and use it in [the Jenkins Swarm plugin](https://github.com/jenkinsci/swarm-plugin/blob/a82051e56baf5f89afe6eb7ed28d02e9fc7eaed8/client/pom.xml#L165-L169). Maybe one day the Jenkins core `ProcessTree` class could be implemented with OSHI and the Jenkins project could get out of the business of writing cross-platform process handling logic. Right now OSHI is better in some regards and `ProcessTree` is better in other regards. OSHI's macOS implementation is clearly ahead of `ProcessTree`, as demonstrated by this PR. 😊 But `ProcessTree` has a better API for getting command-line arguments I think. And the Solaris implementation in `ProcessTree` is more robust than the one in OSHI (which, for example, can only display the first 80 characters of a command line).

**Comment by dbwiddis** [comment]
I missed the subtlety that you only have the `jna` dependency and not `jna-platform`.  `NativeLong` is reasonable.  I'd personally still add some comments documenting why that choice was made should other people see the code and try to do a similar implementation.

Thanks for the compliment! I'll have to take a look at your Solaris implementation.  Although given how long some Solaris bugs have lasted in OSHI, I'm not sure anyone actually uses that port! 🤣

**Comment by ngg** [comment]
The native `long` and `size_t` are always the same size on any reasonable Unix-like system.
Currently Jenkins does not support any of the 32-bit Darwin-based system (x86 macOS or armv7 iOS for example), but IMHO it's still better to use `NativeLong` than `long`.

File: core/src/main/java/hudson/util/jna/GNUCLibrary.java
Comment (basil) [inline core/src/main/java/hudson/util/jna/GNUCLibrary.java:103]: @ngg Could you please add some comments explaining that the use of `NativeLongByReference` and `NativeLong` relies on an implementation detail that is always true in practice and should be replaced with `size_t` and `size_t.ByReference` when we upgrade to JNA 5.9.0 and add `jna-platform` to the build?

**Review by basil — APPROVED** [review]

**Comment by car-roll** [comment]
late to the party here
@timja I'm surprised it takes 15 minutes to actually pass. I would have thought the test would time out before that. I tried to run the tests myself (`Big Sur amd64`), but my tests were still failing. I'm not sure if my environment was acting up though, because at one point I thought I had got something to pass rather quickly 😕 

side note: You only need to run the `NATIVE` tests since the other parameterized tests are all linux containers which wouldn't be affected by this change (most likely environment as to why they were failing).

**Comment by timja** [comment]
@car-roll i had to disable the test timeout to get it to pass

File: core/src/main/java/hudson/util/jna/GNUCLibrary.java
Comment (ngg) [reply core/src/main/java/hudson/util/jna/GNUCLibrary.java:108]: I've added a comment about this.

**Review by timja — APPROVED** [review]

**Review by daniel-beck — APPROVED** [review]
Looks reasonable. Thanks for the detailed explanation.

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/hudson/util/ProcessTree.java
Comment (res0nance) [inline core/src/main/java/hudson/util/ProcessTree.java:1795]: Copying my comment from the other PR:
```
If we respect the length, there doesn't seem to be a reason to clear this.
```

File: core/src/main/java/hudson/util/ProcessTree.java
Comment (res0nance) [reply core/src/main/java/hudson/util/ProcessTree.java:1783]: Copying a comment from another PR:
```
Maybe this makes more sense to be at the start so we don't bother recreating the output stream if we exceed the length. strings should be null terminated so this shouldn't be an issue
```

**Review by oleg-nenashev — APPROVED** [review]
Looks good after going through the issues. It would be really nice to finally have some test coverage for Darwin in CI, but it is definitely a wider topic than this PR. Not blocking, I will go ahead and merge it.

@ngg is there a potential compatibility impact for older platforms? It does not seem so, but just in case

File: core/src/main/java/hudson/util/ProcessTree.java
Comment (ngg) [reply core/src/main/java/hudson/util/ProcessTree.java:1795]: I just added this as another layer of security, yes it's a bit redundant to the length checks, but just to make sure nothing is based on uninitialized memory.

**Comment by ngg** [comment]
> @ngg is there a potential compatibility impact for older platforms? It does not seem so, but just in case

I've been testing this change in our company's CI system for a few weeks, we have Jenkins agents with OS X 10.10 Yosemite, OS X 10.11 El Capitan, macOS 10.12 Sierra, macOS 10.13 High Sierra, macOS 10.14 Mojave, macOS 10.15 Catalina and both x64 and Apple Silicon (M1) based macOS 11 Big Sur. All of these seem to work correctly.

**Comment by oleg-nenashev** [comment]
Thanks! It is ready to go once @res0nance confirms he is fine with the provided responses. I would have merged it to the today's weekly, but our ci.jenkins.io instance has ongoing outage

**Comment by timja** [comment]
> Thanks! It is ready to go once @res0nance confirms he is fine with the provided responses. I would have merged it to the today's weekly, but our ci.jenkins.io instance has ongoing outage

ci.jenkins.io is fine?

**Review by res0nance — APPROVED** [review]

**Comment by dbwiddis** [comment]
Just as an FYI, the `size_t` = `NativeLong` assumption also occurs in the [`pread()`](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/util/jna/GNUCLibrary.java#L83) implementation in GNUCLibrary.  If/when you eventually change this mapping.

**Pull request merged**
